### ISSUE TITLE
False-positive E0:601 on `datetime` when USE_TYPECHECKING

### ISSUE THREAD

**Issue by Kein** [pylint-dev/pylint#9493]
### Bug description

```python
from __future__ import annotations
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from datetime import datetime

def _do_thing(dob: str) -> bool:
    dateBirth = datetime.strptime(dob, "%d.%m.%Y")
    return True
```

### Configuration
```
--disable=broad-except, fixme, unsupported-binary-operation, multiple-imports, global-statement, relative-beyond-top-level,superfluous-parens,import-error,missing-function-docstring,invalid-name,line-too-long,missing-class-docstring,unused-variable,unused-argument,multiple-statements"
```

### Command used

```shell
N/A
```

### Pylint output

```json
[{
	"resource": "/d:/Unlinked//1.py",
	"owner": "_generated_diagnostic_collection_name_#11",
	"code": {
		"value": "E0601:used-before-assignment",
		"target": {
			"$mid": 1,
			"path": "/en/latest/user_guide/messages/error/used-before-assignment.html",
			"scheme": "https",
			"authority": "pylint.readthedocs.io"
		}
	},
	"severity": 8,
	"message": "Using variable 'datetime' before assignment",
	"source": "Pylint",
	"startLineNumber": 7,
	"startColumn": 17,
	"endLineNumber": 7,
	"endColumn": 25
}]
```

### Expected behavior

No error message.

### Pylint version

**pylint 3.1.0:** https://github.com/microsoft/vscode-pylint

### OS / Environment

Windows 10

**Comment by jacobtylerwalls** [comment]
Thanks for the report, but this is intentional. If I add a call to that function at the end of your script, a `NameError` is raised:
```py
% python3 a.py
Traceback (most recent call last):
  File "/Users/jwalls/a.py", line 10, in <module>
    _do_thing('')
  File "/Users/jwalls/a.py", line 7, in _do_thing
    dateBirth = datetime.strptime(dob, "%d.%m.%Y")
                ^^^^^^^^
NameError: name 'datetime' is not defined. Did you forget to import 'datetime'?
```
